lib/streamlit/watcher/LocalSourcesWatcher.py

```diff
@@ -94,10 +94,21 @@ def on_file_changed(self, filepath):
             LOGGER.error("Received event for non-watched file", filepath)
             return
 
-        wm = self._watched_modules[filepath]
-
-        if wm.module_name is not None and wm.module_name in sys.modules:
-            del sys.modules[wm.module_name]
+        # Workaround:
+        # Delete all imported modules so we can guarantee changes to the
+        # updated module are reflected on reload.
+        #
+        # In principle, for reloading a given module, we only need to unload 
+        # the module itself and all of the modules which import it (directly
+        # or indirectly) such that when we exec the application code, the 
+        # changes are reloaded and reflected in the running application.
+        #
+        # However, determining all import paths for a given loaded module is
+        # non-trivial, and so as a workaround we simply unload all watched
+        # modules.
+        for wm in self._watched_modules.values():
+            if wm.module_name is not None and wm.module_name in sys.modules:
+                del sys.modules[wm.module_name]
```

Comment:
That's a really slick solution!

Since @tconkling was already looking at this issue, I'll let him take a look before we merge it in.